src/python/pants/core/util_rules/adhoc_process_support.py

```diff
@@ -542,6 +548,53 @@ async def run_adhoc_process(
     return AdhocProcessResult(result, adjusted)
 
 
+# Compute a stable bytes value for a `PathMetadata` consisting of the values to be hashed.
+# Access time is not included to avoid having mere access to a file invalidating an execution.
+def _path_metadata_to_bytes(m: PathMetadata | None) -> bytes:
+    if m is None:
+        return b""
+
+    def dt_fmt(dt: datetime | None) -> str | None:
+        if dt is not None:
+            return dt.isoformat()
+        return None
+
+    d = {
+        "path": m.path,
+        "kind": str(m.kind),
+        "length": m.length,
+        "is_executable": m.is_executable,
+        "unix_mode": m.unix_mode,
+        "created": dt_fmt(m.created),
+        "modified": dt_fmt(m.modified),
+        "symlink_target": m.symlink_target,
+    }
+
+    return json.dumps(d, sort_keys=True).encode()
+
+
+async def compute_workspace_invalidation_hash(path_globs: PathGlobs) -> str:
+    raw_paths = await Get(Paths, PathGlobs, path_globs)
+    paths = sorted([*raw_paths.files, *raw_paths.dirs])
+    metadata_results = await MultiGet(
+        Get(PathMetadataResult, PathMetadataRequest(path)) for path in paths
+    )
+
+    # Compute a stable hash of all of the metadatas since the hash value should be stable
+    # when used outside the process (for example, in the cache). (The `__hash__` dunder method
+    # computes an unstable hash which can and does vary across different process invocations.)
+    #
+    # While it could be more of an intellectual correctness point than a necessity, tt does matter,
+    # however, for a single user to see the same behavior across process invocations if pantsd restarts.
+    #
+    # Note: This could probbaly use a non-cryptographic hash (e.g., Murmur), but that would require
+    # a third party dependency.
+    h = hashlib.sha256()
+    for mr in metadata_results:
```

Comment:
Is it necessary to dump out the bytes, hash them, and then throw the bytes away? Why not hash the values in that dict directly?

Reply:
In terms of optimizing this computation, if we are fine with adding a third party dependency, we could add a package like https://pypi.org/project/mmh3 in order to use the stable Murmur3 hash algorithm. Thoughts?

But I'd say we hold off on optimizing this for now until it is shown to be a problem.

Reply:
No, I still mean sha256, but just not creating a buffer, and instead passing each value's bytes to the hasher on the fly. But yeah, we can wait to do that if it's an actual problem.